### ISSUE TITLE
EventLoop 3.0 Changes

### ISSUE THREAD

**Issue by rib** [rust-windowing/winit#2900]
Considering the changes being made for `run()-> Result<>`, `pump_events`, `run_ondemand`, and addressing inconsistent ordering guarantees for events like `MainEventsCleared` and `RedrawRequested`, as well as removing and renaming several core events, then it feels like a similar level of change as happened for EventLoop 2.0.

The details are a bit scattered though, and although some things have been agreed as part of lengthy issue discussions we still need to track those ideas at least as a reminder as to what is left to do.

I'm going to be lite on details here, but a lot of the background discussion probably happened here https://github.com/rust-windowing/winit/issues/2706 and maybe a few other issues we can link back here too.

- [x] Replace `run() ->!` with `run() -> Result` as our portable run API available on _all_ platforms (https://github.com/rust-windowing/winit/pull/2767)
- [x] Replace `run_return()` with `pump_events()` and `run_ondemand()` (for Windows, macOS, Linux + Android) (https://github.com/rust-windowing/winit/pull/2767)
- [ ] Maybe add `run_noreturn() ->!` extension API for iOS
- [x] Rename `LoopDestroyed` `LoopExiting` (https://github.com/rust-windowing/winit/pull/2975)
- [x] Remove `RedrawEventsCleared` (https://github.com/rust-windowing/winit/pull/2976)
- [x] Rename `MainEventsCleared` -> ~~`AboutToPoll`~~ `AboutToWait` (Conceptually `NewEvents` and `AboutToWait` can be seen as a way to observe the underlying blocking/polling of the event loop - similar to registering an observer with a macOS `RunLoop`) (https://github.com/rust-windowing/winit/pull/2976)
- [x] Document that `RedrawRequested` is no longer guaranteed to be delivered in order with respect to any other event - it is platform specific 
- [x] Make it clearer that applications should redraw any time they get a `RedrawRequested` callback (and they shouldn't defer the redraw outside of the winit callback).
- [ ] Make it clearer that `Resumed` events are delivered on all platforms (not just mobile) and `Resumed` events are a good place to initialize an application because it's the first point where it's possible to start creating Windows across all platforms.
- [x] Remove lifetime from the `Event`
- [ ] Aim to implement `RedrawRequested` throttling by default where there's a suitable mechanism provided by the window system
  - [x] By `WM_PAINT` on Windows
  - [ ] Either with `drawRect` or display link on macOS (same for iOS?)
  - [x] By frame callbacks on Wayland
  - [x] Via requestAnimationFrame for Web
  - [ ] Maybe via Choreographer on Android
  - [ ] By praying on X11

_I don't mind if we want to collate this information somewhere else, or in a different, better way etc but wanted to a least get a basic task list laid out somewhere_

Maybe a milestone would be a nicer way to organize this?

**Comment by rib** [comment]
CC: @kchibisov, @msiglreith, @madsmtm, @daxpedda

**Comment by notgull** [comment]
> * Maybe add `run_noreturn() ->!` extension API for iOS and Web

+1 to this. Maybe it could also be implemented for all of the non-Android platforms, which use `std::process::exit()` instead.

> * Make it clearer that `Resumed` events are delivered on all platforms (not just mobile) and `Resumed` events are a good place to initialize an application because it's the first point where it's possible to start creating Windows across all platforms.

Maybe we build this into the type system? I.e. `Resumed` returns a `WindowCreationToken` or something that can't be created outside of `winit`, that is needed to create windows?

**Comment by kchibisov** [comment]
>+1 to this. Maybe it could also be implemented for all of the non-Android platforms, which use std::process::exit() instead.

This API only required on ios. 

>The only other thing is that RedrawEventsCleared was nice for submitting the wgpu queue for when all redraws are done. It would be nice if something similar were exposed on platforms that support it.

It'll be named `AboutToPoll`.

**Comment by kchibisov** [comment]
Wrt `MainEventsCleared` and `RedrawEventsCleared`, I think `MainEventsCleared` is here to stay with us. The reason for that is that winit has a guarantee right now is that if you do `request_redraw` prior to `MainEventsCleared` including it, it'll be scheduled on this loop iteration.

While throttling adds a bit to that, I think it'll still be nice to have such entry point to potentially schedule the redraw on this loop iteration, so for example when after processing input events, you decided to request a redraw, so if on e.g. Wayland you already had frame callback, you could simply draw on this loop iteration reducing the overall latency.

**Comment by kchibisov** [comment]
Wrt to window creation, we know that it's generally async, maybe we should expose it? Though, I'm thinking that it'll too much. But that, however will make it work with `Resumed` without a type system level token and such, you could request a window outside of the loop, but you'll get it only once you start the loop.

It could be like `Window::request_new() -> Result<WindowId, Err>`, so you're getting a `WindowId` ahead of time for your new window, thus you could have a spot for it in the `HashMap` for example. And then you receive an even with the `Window`, that way you won't have issues with calling `inner_size` way too early, because the window will be still in process, and you won't have an issue with callback stacking if we decide to call from e.g. wayland events being dispatched directly to the user.

But I fill like that could be a bit a **change** in convenience, but generally speaking it'll improve situation wrt correctness, given that you really want to create EGLSurface for Wayland with the `right size` from the start, and not with some arbitrary size you pick, it's possible right now, but users rely on their sizes and not on `inner_size()` calls, and on some systems `inner_size()` is not ready early on.

Besides, macOS has issues in general when it comes to creating windows out side of the loop, so some windowing features stop working (alacritty thus moved to `Resumed` event instead).

**Comment by rib** [comment]
> Wrt to window creation, we know that it's generally async, maybe we should expose it?

I think it could be good to expose that window creation is async and fix the awkward issue with Wayland right now with the sync roundtrip in `Window::new()` - that is having a really disproportionate affect on the whole design of the Wayland backend currently.

I don't know about really buffering requests to create a window right from before an event loop is started. It's always seemed a bit odd to me with Winit that you could create windows before starting to run an EventLoop.

The idea of initially just getting a `WindowId` sounds like it could work.

In terms of sizes to use with EGL there's another big mess imho with the `inner_size` API being overloaded which is a separate issue really.

It wouldn't necessarily always bad for apps to optimistically use a size that was used to request a new window for an egl surface if it e.g. means they don't have to wait for a round trip - if it's 99% likely they get what they asked for and the other 1% will correct itself via resize events etc.

**Comment by rib** [comment]
> The only other thing is that RedrawEventsCleared was nice for submitting the wgpu queue for when all redraws are done

There would no longer be any notion of having multiple redraw events within a single iteration of the loop - you just get (throttled) RedrawRequested events in response to calling request_redraw() - so now you would just submit work to the GPU as part of your RedrawRequested event handler (i.e. after you've finished actually drawing). If we kept a RedrawEventsCleared then conceptually it would just be called back-to-back, immediately after the `RedrawRequested` event - which makes it pretty much redundant.

**Comment by daxpedda** [comment]
I don't think we want a `run_noreturn() ->!` as an extension API for Web if we already have `spawn`.
Is this desirable somehow?

**Comment by kchibisov** [comment]
I also don't quite understand why we have an api which `-> !` if `web` doesn't really need it.

The whole point was to move away of `-> !`, which is basically ios, and ios in my opinion could live in its own with `-> !`, given how niche the platform is for rust development, and that users targeting it likely have a separate path regardless.

The only issue is that I think `spawn` on web doesn't block at all? While the default API we want should block the caller until `run` finishes.

**Comment by daxpedda** [comment]
> The only issue is that I think `spawn` on web doesn't block at all? While the default API we want should block the caller until `run` finishes.

Yes, that unfortunately can't be fixed in Web.

**Comment by kchibisov** [comment]
I think we've discussed in the past and we could have a way to block via the exception handler, basically the way `run -> !` was implemented before.

**Comment by daxpedda** [comment]
> I think we've discussed in the past and we could have a way to block via the exception handler, basically the way `run -> !` was implemented before.

Whats the "exception handler"? Right now `run -> !` on Web is implemented by throwing a panic, which "works".

**Comment by rib** [comment]
it's not a panic that's thrown though is it? I thought it was done by punching through to javascript to throw an exception which Rust wouldn't see

**Comment by rib** [comment]
which was an old hack from emscripten that used to be done to give the illusion of a function in C/C++ that doesn't exit

**Comment by rib** [comment]
> I don't think we want a `run_noreturn() ->!` as an extension API for Web if we already have `spawn`. Is this desirable somehow?

yeah I don't really feel like it's necessary for web.

for web you're going to have a bit of bespoke wasm_bindgen glue anyway so it's just not really an issue to use a different run function

**Comment by rib** [comment]
it's pretty much just iOS that wants run_noreturn so I really don't see that it would be that nice to expose it across all platforms just for the sake of making the iOS behaviour into a lowest common denominator. Having every other platform hit `process::exit()` just feels really yukky to me.

technically though we could implement it for all platforms except Android I think, but that would kind of endorse it more than I feel like it should be.

**Comment by rib** [comment]
I'd guess iOS is probably also like Android and Web where you're very likely going to have a bespoke entrypoint that isn't just a plain `fn main()` so I'd guess it's also a non-issue for iOS to have a special case `run_return() -> !` if they want to make it clearer that the function doesn't return.

**Comment by rib** [comment]
> There isn't a big difference in practice between throwing directly and calling panic! though.

The trick relies on Rust basically not being aware of the Javascript layer throwing the exception - it's transparent to the Rust compiler, so I think it's maybe quite an important distinction - a panic is something that Rust would definitely know about.

**Comment by rib** [comment]
> Maybe we build this into the type system? I.e. `Resumed` returns a `WindowCreationToken` or something that can't be created outside of `winit`, that is needed to create windows?

This idea would potentially work better if there was an event that corresponded to `FirstResume` which doesn't currently exist.

It would be tricky to do with Resumed events that can be emitted multiple times during the life of mobile apps.

It could be interesting to consider since it can be good when the API forces you into a portable pattern - if it doesn't break other  important use cases.

**Comment by daxpedda** [comment]
> > There isn't a big difference in practice between throwing directly and calling panic! though.
> 
> The trick relies on Rust basically not being aware of the Javascript layer throwing the exception - it's transparent to the Rust compiler, so I think it's maybe quite an important distinction - a panic is something that Rust would definitely know about.

Panics on Wasm are implemented very crudely right now. Basically any sort of panic in Wasm just calls the trap instruction, which throws an exception in the browser environment. The Rust compiler doesn't do anything else here.

The important thing is that we don't trigger the panic handler, which a call to `panic!` would, because the user might have registered a custom panic handler, which might have unintended consequences. But the default panic handler and calling panic itself has basically no side effects except throwing an exception right now.

That is all until Rust gains Wasm unwinding support.

**Comment by madsmtm** [comment]
The problem in my eyes with `run()` not returning `!` is that people assume that they can do cleanup / other things after the event loop is done, while in reality they cannot do that portably.

Though I didn't know about `process::exit()` being problematic on Android, I guess the downside I'm noting above is the lesser evil.

**Comment by daxpedda** [comment]
> The problem in my eyes with `run()` not returning `!` is that people assume that they can do cleanup / other things after the event loop is done, while in reality they cannot do that portably.

That is a good point ... I didn't consider that.
Honestly I don't think we should change the default `run() -> !` to `run() -> Result` then, because that's gonna be fundamentally incompatible with Wasm.

I think @kchibisov mentioned in the PR that we should consider removing `run() -> !` for Wasm, which now sounds quite a bit better then offering `run() -> Result` in Wasm at all.

I am strongly in favor of not making the change to `run() -> Result`, it won't be cross-platform.

**Comment by rib** [comment]
> I am strongly in favor of not making the change to run() -> Result, it won't be cross-platform.

:( we've been around in circles so many times on this now.

We can't support `run() ->!` on Android because it's not acceptable to just kill the whole process as a way to not return and Rust has no API to just terminate a thread.

`run() -> Result` is the more cross-platform API.

Overall there's no perfect `run()` signature for all platforms.

Another option could be to rely on extensions for _all_ platforms since there's no perfect solution but as it it `run() -> Result` can actually be supported on all platforms so we don't need to do that.

The ability to return a status is something that _is_ really desirable and also possible on the majority of platforms except for iOS and Web.

It's notable that `run() ->Result` API doesn't need to be used on Web or iOS and since it's very common (esp for Web) to have extra glue for the entrypoint for the app it also seems pretty reasonable that for web you could use a special `run_noreturn` to better communicate to the Rust compiler that `run()` doesn't return.

 It doesn't break Web or iOS for the compiler to not inform apps that the function doesn't return - that's an ergonomic hint and we can provide an extension API like `run_noreturn` to give you the ergonomics you want on Web and iOS.

**Comment by daxpedda** [comment]
> :( we've been around in circles so many times on this now.

Apologies, I wasn't here for that. I'm still new here!

> We can't support `run() ->!` on Android because it's not acceptable to just kill the whole process as a way to not return and Rust has no API to just terminate a thread.

Yeah, that's a problem, I agree.

> Another option could be to rely on extensions for _all_ platforms since there's no perfect solution but as it is `run() -> Result` can actually be supported on all platforms so we don't need to do that.

I think this is the most reasonable option here.
As far as I can tell there is no one cross-platform API that fits the bill here. `run() -> Result` is not acceptable for API, this is a big footgun and the fact that this doesn't return is not reflected by the API, it just doesn't make any sense to me.

---

I would be in favor of just making this platform specific, or alternatively, just expose `run() -> Result` on all platforms except Web.

**Comment by kchibisov** [comment]
>I think this is the most reasonable option here.
As far as I can tell there is no one cross-platform API that fits the bill here. run() -> Result is not acceptable for API, this is a big footgun and the fact that this doesn't return is not reflected by the API, it just doesn't make any sense to me.

moving to extensions is fine, I guess.

**Comment by rib** [comment]
I tend to think that the fears around `run() -> Result` are way over blown but yeah, just moving to extensions also sounds OK  to me too.

Especially considering the extra steps to create the bindgen entrypoint for a Web app I just see it as mostly academic.

It's similar to how the C API for NSApplication technically returns in `int` but actually never returns - it just doesn't matter much except for ergonomics and helping hint to developers that subsequent code wont run.

Since the `run() -> Result` change is already made in the pump_events branch I'd prefer to initially land that as is initially but can follow up with making `run() -> Result` an extension

**Comment by notgull** [comment]
I'm wondering if we should adopt the way that `sokol` does things and just control the entry function in code. Combined with #2903, it would look something like this:

```rust
#[winit::main]
fn main() -> impl Application {
    // return an object with callback handlers
}
```

This would then expand to whatever is needed to handle the weird event loop returns on Web, Android, etc etc.

**Comment by kchibisov** [comment]
I think such main entry point is for the gui toolkits, and we try to provide building blocks for such toolkits. I know that for example applications where they plug winit as yet other event source and target only Wayland/X11 won't be happy with things like that.

**Comment by daxpedda** [comment]
I we intend to make run functions platform specific, I would prefer to not hide it behind extension trait because it's shared by so many platforms otherwise. But instead only hide the methods behind `cfg` guards.

**Comment by rib** [comment]
> This would then expand to whatever is needed to handle the weird event loop returns on Web, Android, etc etc.

On Android the `android_main` API is passed on app which both the application an/or Winit would need to create an event loop which would be awkward to paper over with this kind of `main()` macro.

I tend to think these kinds of magic macros are more appropriate for higher-level frameworks that can be more opinionated, since they can also do things like setting up logging which may involve different crates/APIs for different platforms.

**Comment by madsmtm** [comment]
I'll say it again, I'm fine with `run() -> Result<(), E>` on iOS, and then let it be a bit footgun what code you can run afterwards (as long as we document that).

**Comment by parasyte** [comment]
I need to spend more time going over this discussion, but I would also like to make an alternative proposal. I have concerns that the inversion of control used by the event loop is broadly responsible for many issues around managing window lifetimes. And because I haven't seen this alternative mentioned on any of the linked topics, I think it is prudent to bring it up now.

One prominent example of an issue caused by this event loop design is https://github.com/gfx-rs/wgpu/issues/1463, and more rationale is provided in the documentation for the [`winit-main`](https://crates.io/crates/winit-main) crate. Which (not surprisingly) reverts the control, giving it back to the caller. This crate is grossly outdated, but it's generally a good idea.

[`async-winit`](https://crates.io/crates/async-winit) is another attempt to revert control. This one is closer to what I would expect to be provided by `winit` itself. The design of this crate forces the user to collect events from disparate sources into a single stream to avoid issues with mutable borrows that must be used in response to different events. (For instance, [`pixels`](https://github.com/parasyte/pixels) needs to respond to `Resize` and `RequestRedraw` events by updating internal state. And those methods take `&mut Self` references.)

Yet another attempt is [`winit-block-on`](https://docs.rs/winit-block-on), which was at least mentioned in https://github.com/rust-windowing/winit/issues/2431#issuecomment-1455941076. I don't have any experience with this crate to comment.

Moreover, having `run_return` and `run_ondemand` _at all_ is a strong indication that the inversion of control is itself the problem.

My preferred event loop might look like this [^1]:

```rust
// The `StreamExt` trait makes our event loop comfortable to use.
use futures::StreamExt as _;

fn main() {
    // The async block is spawned on a runtime running off the main thread.
    // Events are passed through the `EventLoop`, which the async block processes.
    winit::run(async {
        let event_loop = EventLoop::new();
        let mut window = None;

        while let Some(event) = event_loop.next().await {
            match event {
                Event::Resumed => {
                    window = Some(Window::new().unwrap());
                },

                Event::WindowEvent {
                    event: WindowEvent::CloseRequested,
                    window_id,
                } if window_id == window.unwrap().id() => {
                    break;
                },
            }
        }

        // It's OK to either return from the async block (i.e. exit the app) or
        // create an additional loop to continue consuming from the stream.
    });
}
```

# Evaluation and analysis

How does this alternative proposal address https://github.com/gfx-rs/wgpu/issues/1463?

Because this API does not invert control with a callback (in principle), there is no need to move the window into the closure (or share any state with a callback for any reason). That means the window can stay on the stack [^2]. And ultimately the window can be borrowed trivially by libraries that use `raw_window_handle` to abstract away the underlying window provider. While also allowing the event loop [^3] to call methods on the window. No need to mess around with reference counting or leaking a `Window` to satisfy [the `'static` lifetime requirement](https://docs.rs/winit/latest/winit/event_loop/struct.EventLoop.html#method.run).

As a proof of concept, I created a branch in the `pixels` repo that adds a lifetime parameter needed to resolve the wgpu soundness issue. It's largely incompatible with the way `winit` forces inversion of control with a callback, but works well enough with `async_winit` (and the code isn't a terrible departure from the ideal). The POC is very incomplete, but you can see it here: https://github.com/parasyte/pixels/commit/932507818812c417787ce80aa579a43ba0462ba9

I am convinced it's worth investigating further in this direction, if not outright investing in it now [^4]. I basically want to throw a spanner into the works here and push for not just an incrementally better event loop design, but a radically better event loop design. And something like this that reverts control back to the user is a good starting point.

# Caveat

There is one caveat that I am aware of, and that's properly handling events which cannot be queued. See "portability hazards" in https://github.com/rust-windowing/winit/issues/2706#issuecomment-1462423107. It seems to me this shouldn't really be much of an issue, since the main thread can always block on events that are handled in another thread without queueing (cooperative multithreading). It would just require designing the `Event` type to allow notifying the main thread when the event has been handled. Here's a silly example to sketch the idea:

```rust
while let Some(event) = event_loop.next().await {
    match event {
        Event::RedrawRequested(notifier, _) => {
            // Do drawing stuff here, which may take non-trivial time...
            do_draw();

            // Notify the main thread when drawing is complete.
            // Or perhaps the main thread could be notified automatically when `Event` is dropped?
            notifier.notify().await;
        }

        // ...
    }
}
```

[^1]: Yes, I realize that this `Stream`-based API is still "inversion of control" proper. But like all good uses of `Future`s, it _looks_ and _feels_ procedural, and that's the whole point!

[^2]: Technically it's owned by the `Future`. But for the sake of argument, let's just assume it's on the stack.

[^3]: The user's event loop, not `EventLoop`... Which I was tempted to rename to `EventStream` or something else in the "ideal" example.

[^4]: I have other motivations for "reverting control" to the app. And I intend to write more in depth on the subject when I can get my thoughts together better. But for some additional context, I did have this same problem with a completely different product written in a different language (JavaScript). Turning the callback-based API into an async streaming API was beneficial, again for most of the same reasons: https://github.com/parasyte/async-pubnub-js

**Comment by kchibisov** [comment]
>One prominent example of an issue caused by this event loop design is gfx-rs/wgpu#1463. More rationale for not inverting control is provided in the documentation for the winit-main crate. Which (not surprisingly) reverts the control, giving it back to the caller. This crate is grossly outdated, but it's generally a good idea.

This issue has nothing to do with the loop design of the winit, it has to do with the poor API around the Window, window must be created only via the event loop and tied to its lifetime
and delivered back via the event. This is how it should work and this is how it works internally, yet we force sync-ness on the async window creation by simply doing `.await`.

>async-winit is another attempt to revert control. This one is closer to what I would expect to be provided by winit itself. The design of this crate forces the user to collect events from disparate sources into a single stream to avoid issues with mutable borrows that must be used in response to different events. (For instance, pixels needs to respond to Resize and RequestRedraw events by updating internal state. And those methods take &mut Self references.)

Fencing is the way it should be done and I've already suggested fencing in some other issues with multi-threaded syncing issues.

The main issue with async is that winit must write its own executor, as in, its event loop must become an executor. None of the existing executors will help us here unfortunately due to how it all works on macOS/ios, it's also not really async on some platforms and we can't run UI tasks off the main thread (all must be concurrent).

Maybe @notgull has an idea how to improve the situation here or approach what you suggest, but for now the only thing I see is the poor man's async based on event loop of the winit.

But generally speaking, I don't like the proposed loop, nor do I like the current loop we have, I'd like to have real callback based API, so we could pass specialized objects for e.g. keyboard, ime, etc, right for corresponding callbacks. And generally make an API trait based, because that's how it really should be.

The issue with the proposed API is only lifetime of the window, but once we make window creation basically an event you get from the event loop, it'll be hard to move it away. I'm not so sure how we could properly define a lifetime when we really want an ability to send a window across the threads (unless we make all the threads scoped and tie their lifetime to event loop's one).

**Comment by parasyte** [comment]
> This issue has nothing to do with the loop design of the winit

It most certainly does. The existing callback for the `run` method can only contain `'static` lifetimes. The window must be owned outside of the closure, and anything that references it can only do so with a `'static` lifetime (`Arc`, leaking, etc) or it must take ownership of the window. This is all described in the wgpu thread, but the root cause is winit's callback design.

> The main issue with async is that winit must write its own executor, as in, its event loop must become an executor.

The only reason to use async at all is because it is required by the web platform. Otherwise, I would prefer to consume events from a `std::sync::mpsc`.

> None of the existing executors will help us here unfortunately due to how it all works on macOS/ios, it's also not really async on some platforms and we can't run UI tasks off the main thread (all must be concurrent).

This seems like the biggest issue, but I covered it in the "caveat" section above. The async executor runs on its own thread and the main thread communicates with it by message passing. The main thread can block until the event has been handled; the handler sends a notification to wake the main thread. It doesn't look like there is really anything limiting the handler to _running on the main thread directly_. Pretty much everything you need for drawing is `Send` and `Sync`! The `Window` reference, the wgpu `Surface`... [^1]

> But generally speaking, I don't like the proposed loop, nor do I like the current loop we have, I'd like to have real callback based API, so we could pass specialized objects for e.g. keyboard, ime, etc, right for corresponding callbacks. And generally make an API trait based, because that's how it really should be.

That is an opinion. My opinion is that callbacks are all wrong. 2023 could be the year of callback hell.

> The issue with the proposed API is only lifetime of the window, but once we make window creation basically an event you get from the event loop, it'll be hard to move it away. I'm not so sure how we could properly define a lifetime when we really want an ability to send a window across the threads (unless we make all the threads scoped and tie their lifetime to event loop's one).

I'm not sure what you mean by the the proposal having lifetime issues. It solves lifetime issues.

[^1]: The softbuffer `Surface` is `!Send` and `!Sync`. Might want to fix that! But the surface can be created off the main thread because the `Window` is `Send` and `Sync`!

**Comment by kchibisov** [comment]
>It most certainly does. The existing callback for the run method can only contain 'static lifetimes. The window must be owned outside of the closure, and anything that references it can only do so with a 'static lifetime (Arc, leaking, etc) or it must take ownership of the window. This is all described in the wgpu thread, but the root cause is winit's callback design.

Or by the event loop itself, so the way to access could be via the method on event loop window target.

>I'm not sure what you mean by the the proposal having lifetime issues. It solves lifetime issues.

I mean that your window is a property of the event loop, it's true that the current design could have issues with the wgpu, but have you seen https://github.com/rust-windowing/raw-window-handle/pull/132 release adding safe raw window handles? They all creating your stuff safely, and they work even with `glutin` just fine, as well as many other crates. They do handle Android, and so on.

>That is an opinion. My opinion is that callbacks are all wrong. 2023 could be the year of callback hell.

Having 2 threads for some basic windowing is a complete non-sense to me, sorry. You don't need that many resources for really no reason. Be aware that half of the windowing/gl context API on macOS/ios **must be** called from the main thread, otherwise it'll SIGTRAP. If you really want async **the way you describe**, you can use async winit with your proposed design the channel, don't see a reason to have it inside the winit repo.

If the async will be in the winit, it must use the proper scheduling without adding overhead of extra threads and inner channels. The only thing the current winit design is lacking is a proper callback machinery where you have zero inner buffering on **all** platforms between getting an event and handling it.

You can't send neither EGLSurface nor softbuffer surfaces across the threads, they must live on the thread they were created, if they're moved it could segfault. With EGL for example the only thing you can send is the not current context.

**Comment by notgull** [comment]
> Maybe @notgull has an idea how to improve the situation here or approach what you suggest, but for now the only thing I see is the poor man's async based on event loop of the winit.

Disclaimer: I'm very biased in one direction of this debate here. I'm the author of two of the "control inversion" abstractions mentioned above (`winit-block-on` and `async-winit`). In addition, I'm one of the primary maintainers of `smol`. Needless to say, I'm a big fan of the `async` style of programming.

The problem comes down to the fact that all of the operating systems expose very different ways of handling events. Some of them are only barely compatible with eachother.

- Wayland and X11 are basically just polling a socket, so the sky's the limit with them. You can fit just about any pattern into here; hell, you can even integrate these into external event loops if you wanted to!
- Windows has the weird class-based message queue. You can poll for events in a blocking or non-blocking way, but once you `Dispatch` them you have to deal with them in callback form; often times multiple at once in unpredictable and re-entrant ways. Unfortunately there's no way to register a `Waker` to tell if the current thread has any events ready, so it's difficult to integrate into external loops.
- macOS and iOS use Apple's event loop abstraction that *really* likes being in complete control of the entire program, and also has the bonkers main thread limitation. This one is pretty much the lowest common denominator here, since it's hard to mold it into any other patterns. Apple is pretty much holding us hostage here.
- Android is basically just callbacks all the way down, but at least you can control *when* you poll for events. I'm not too familiar with the underlying infrastructure here but to me it seems similar to how Windows works. There's also the restriction where `std::process::exit()` can't really be called because we use Activities instead of straight processes.
- Web needs a non-blocking callback-based loop. I'd argue this is similar to the Apple loop, at least from a superficial level.

It seems to me that the greatest number of systems have a loop that needs complete control over the main thread, and that events are generally dispatched directly from the windows and the devices rather than the other way around.

Therefore I think that either @kchibisov's callback based model or my own `async` model. I'm biased in the direction of my own model, since I think it fits Rust's general flow better. However I know there are already many projects that take advantage of the callback-based model. Either way works; it's easy enough to build an `async` layer on top of callbacks and vice versa.

**Comment by parasyte** [comment]
The real issue appears to be that you are arguing from a position of keeping the event loop as close to all supported platforms as possible. I'm arguing from the opposite direction; providing an abstraction that correctly encapsulates the incompatibilities between platforms.

If there is a hard constraint to provide the lowest possible overhead, then the API is always going to suffer from the problems that are trying to be addressed by EventLoop 3.0. The least common denominator (callbacks) will limit safety and expression in application architecture.

Obviously additional abstractions can always be built on top, but that goes for every level of abstraction. Async is not ideal for desktop applications, it's just the least common denominator abstraction that makes sense because the web platform requires it (not allowed to block the main thread). Even if an async event loop was provided, I would still like a blocking abstraction on top of it because it's a better fit for everything other than web.

So, I can't quite buy the suggestion that building an async abstraction on top of callbacks is the best approach for something so fundamental. We'll end up in a situation where `winit` is no longer the crate that developers use to abstract over platforms for GUIs. It will be a dependency of the crate with a usable API.

**Comment by kchibisov** [comment]
>The real issue appears to be that you are arguing from a position of keeping the event loop as close to all supported platforms as possible. I'm arguing from the opposite direction; providing an abstraction that correctly encapsulates the incompatibilities between platforms.

That's how winit started with the `poll` API, it wasn't working.

>If there is a hard constraint to provide the lowest possible overhead, then the API is always going to suffer from the problems that are trying to be addressed by EventLoop 3.0. The least common denominator (callbacks) will limit safety and expression in application architecture.

Callbacks are way safer if you cook them correctly and with callbacks we can enforce sync handling of various stuff. The issue is purely with the current design.

>Async is not ideal for desktop applications, it's just the least common denominator abstraction that makes sense because the web platform requires it (not allowed to block the main thread).

You can build async on top of regular callback model, in fact winit is **async**, we just don't use the keyword and our executor is an event loop itself, because some requests on a window are async and you get something out of them back.

I'm not against `async` in particular, it's just I don't see how we can make it work in a sensible way. 

>Even if an async event loop was provided, I would still like a blocking abstraction on top of it because it's a better fit for everything other than web.

I also want something like that with the current async model of winit, but the problem here, is when you try to wait for request to complete you get all other events along what you've been waiting for from the desktop system. So, in your words, once you do `.await` on e.g. some window request, you could get back something from the event loop that must be **synced** and if you don't sync to it, you could get a `protocol error` on e.g. Wayland and straight crash. All events, in gerenal, must be processed sequentially in order they arrive to the app.

>So, I can't quite buy the suggestion that building an async abstraction on top of callbacks is the best approach for something so fundamental. We'll end up in a situation where winit is no longer the crate that developers use to abstract over platforms for GUIs

The current winit's callback design is pretty poor. However the situation will get improved, it's just this particular issue has nothing to do with how we change/improve the callbacks based machinery.

Also, coming to your original issue with `wgpu`, whatever winit does has nothing to do with the `wgpu`, because `wgpu` must not assume that it's using winit, so the only way to make anything safe is the `safe window handles` initiative lead by @notgull .

**Comment by parasyte** [comment]
> All events, in gerenal, must be processed sequentially in order they arrive to the app.

This is precisely why I had to "reinvent" the `Event` enum in my PoC with `async-winit`.

> Also, coming to your original issue with `wgpu`, whatever winit does has nothing to do with the `wgpu`, because `wgpu` must not assume that it's using winit, so the only way to make anything safe is the `safe window handles` initiative lead by @notgull .

`wgpu` has no choice to make its API safe without taking ownership (or requiring shared ownership) over the window because the `winit` event loop callback does not allow temporary borrows to anything in the outer scope. The callback approach makes temporary borrows impossible as long as the function has a `'static` lifetime constraint on the environment it captures. As I understand it, this constraint comes from the web backend: https://github.com/rust-windowing/winit/issues/1543#issuecomment-616833377. It doesn't matter how much `safe window handles` wants to help, this is a problem with the `'static` lifetime on the callback.

**Comment by parasyte** [comment]
I'm not sure why we're talking past each other.

There isn't any reason to focus on `wgpu` when the example in #1543 is an adequate description of the problem that `wgpu` has (and that example clearly has nothing to do with `wgpu`). Bringing up `wgpu` again is ignoring the elephant in the room and feels like an attempt to redirect the discussion.

Using `Arc` and `Rc` fall under the shared ownership option, which I've already mentioned.

**Comment by notgull** [comment]
Regarding `wgpu`, I suppose that I should link https://github.com/gfx-rs/wgpu/pull/3932 here

**Comment by rib** [comment]
@kchibisov for the rename of `MainEventsCleared` I'm wondering if `AboutToWait` might be slightly better than `AboutToPoll` because there's no other public "poll" concept in the API for Winit and "poll" in this context is also effectively unix-specific terminology.

More generally in computer science a "poll" would be a non-blocking check operation but that's almost the opposite of what it means here and with the unix poll apis.

**Comment by kchibisov** [comment]
yeah, sounds fine.

**Comment by rib** [comment]
Catching up with some of the comments above about inversion of control, I think more generally speaking it's probably be best to open new issues / discussions for airing bigger design ideas for the event loop.

I wouldn't worry about needing to air big ideas in case we miss our chance to implement them here. I'm sure these changes won't be the last time we significantly iterate on Winit's event loop APIS - I can certainly see lots of room for further iteration, and can easily imagine wanting an "EventLoop 4.0" at some point too.

The general idea of inverting control comes up pretty regularly, as well as various discussions around supporting async better and they generally raises many of the same portability challenges that come from balancing a pretty complex matrix of constraints to try and support lots of platforms with some very significant architectural differences. E.g. here's one recent discussion: https://github.com/rust-windowing/winit/issues/2857

There's certainly room to experiment though with continuing to iterate the design of Winit event loop and e.g. being able to marry it with Rust's async model in some way but there are lots of details to work through here.

I think it's one of those things where it's also perhaps relatively easy to tackle for 90% of the platforms but potentially harder to come up with a robust design that works for all platfoms (e.g. considering general lack of threading on Web, and platforms based on callbacks and callbacks that need to be synchronized with the OS).

In the short-term (like hopefully for 0.29) I'm hoping to keep this issue generally focused on tracking the more-specific changes that we currently have in-flight for iterating on Winit's current event loop API.

**Comment by parasyte** [comment]
While not an outright dismissal, the lack of enthusiasm for accepting new ideas is concerning.

**Comment by kchibisov** [comment]
@parasyte I mean, we don't have that much resources to evaluate every idea especially when it's being suggested right in the end of the discussion here.

Every one is free to prototype the design and propose it, so we can discuss. Prototyping is essential because some backend may simply not work with the proposed changes and it's better.

Winit can't do radical changes, because there are a lot of big projects depending on it, so every approach should be incremental and provide a justification compared to the current design.

Right now I don't see that, and I'm more interested into changing it into into a trait based system, because that's an incremental step forward and will allow to handle your particular issue and issues  related to the application life cycle.

**Comment by rib** [comment]
> While not an outright dismissal, the lack of enthusiasm for accepting new ideas is concerning.

@parasyte , that really was not my intent at all, sorry if that's how it came across. I intentionally tried to invite the ideas, but at the same time try to ensure that this specific issue could remain focused on tracking the changes that are already in progress.

> "I can certainly see lots of room for further iteration, and can easily imagine wanting an "EventLoop 4.0" at some point too."

> "There's certainly room to experiment though with continuing to iterate the design of Winit event loop "

I honestly was trying to balance my wording to try and avoid coming across like I was shutting down your ideas.

I was also trying to reassure that there's no urgency here to have to evaluate alternative, significant event loop re-designs here.

Personally I can see lots of different places where Winit's design ideally needs iteration and anyone should feel free to explore ways in which it can be improved.

**Comment by parasyte** [comment]
@rib Your response was rather diplomatic, I am more concerned with how others reacted.

Also, I am obviously out of the loop. From my perspective, I was providing feedback quite early in the change cycle. But it sounds like that isn't accurate, and it's actually late in the development cycle? This ticket had been open less than 4 weeks when I initially commented. Surely my confusion is not unfounded?

**Comment by kchibisov** [comment]
@parasyte the ticket was opened as a follow up to the https://github.com/rust-windowing/winit/pull/2767 (Apr 12) and we've discussed it in Feb-March for example. The idea to open is to track the updates discussed back then.

**Comment by rib** [comment]
Yeah unfortunately the full background for this tracker issue probably wasn't entirely obvious without the context from having followed the discussion in #2706 and review process for #2767.

The original #2706 discussion ended up covering quite a lot of ground and we came up with a relatively specific plan for a set of changes that would help address an number of specific event loop bugs/issues that mostly related to `run_return` shortcomings.

I ended up opening this issue because the details were getting a bit fragmented - e.g. with side discussion on IRC/matrix, plus losing track of the details from #2706 and so I wanted to have a central reminder of the things that had been agreed - and to also help double check that we were one the same page about that, while I was working on the implementation.

So really by the time I opened this issue, I had actually already implemented a big chunk of that plan.

I ended up retrospectively referring to the set of changes as "EventLoop 3.0" just because in totality the scope of what we had ended up agreeing via #2706 / #2767 turned out to be very similar in scope to the changes that were made for "EventLoop 2.0".

**Comment by notgull** [comment]
We should also decide on a middleware pattern for the event loop. There is a decent batch of crates that "intercept" events from the event loop for some extra functionality. Off the top of my head there is [`winit_input_helper`](https://crates.io/crates/winit_input_helper) and [`accesskit-winit`](https://crates.io/crates/accesskit_winit). It would be nice if we standardized a pattern for this like [`wgpu` did](https://github.com/gfx-rs/wgpu/wiki/Encapsulating-Graphics-Work).

My personal vote would be for a pattern like this:

```rust
impl FnMut(&mut EventLoopWindowTarget<T>, Event<T>) -> KeepGoing<T>
```

Where `KeepGoing` is an enum with `Intercepted` or `Continue(Event<T>)` variants, to indicate whether the event was consumed or not.

This would be nice to be able to easily integrate these crates into `async-winit`, since that crate could just expose an `add_middleware` function or something to integrate those crates in.

**Comment by nerditation** [comment]
> We should also decide on a middleware pattern for the event loop. There is a decent batch of crates that "intercept" events from the event loop for some extra functionality.

can I take this chance of EventLoop 3.0 discussion to ask a question about `EventLoopWindowTarget`, is there's a reason it need to be generic other than it's evoled from the implementation? I think only `EventLoop` and `EventLoopProxy` actually needs `UserEvent`, but `EventLoopWindowTarget` doesnt, since it's the `Event` type, not the `WindowEvent` type, that is parameterized.

I ask because when I was experimenting with an idea kind of like a *middleware*, e.g. I have a function with a signature like this:

```rust
fn make_handler<T: 'static>(
	self,
	mut user_handler: impl FnMut(Event<T>, &ELWT<T>, &mut ControlFlow),
) -> impl FnMut(Event<Signal<T>>, &ELWT<Signal<T>>, &mut ControlFlow) {
	todo!()
}
```

and I got stuck. the `Event<Signal<T>>` can be easily mapped to `Event<T>` thanks to the `Event::map_nonuser_event()` api, but there's no way to map an `ELWT<T>` type. and I think about it for a while, realizing the `ELWT` concept probably could be reworked (or simply removed), if the event handler type is implemented using a custom trait instead of `FnMut()`, similar to the idea in #2903 

----

btw, I discovered accidentally, my signature is similar to this `winit-block-on` function:

link: https://github.com/notgull/winit-block-on/blob/fc574c10503ed187d53b2f2c7a344fc5dd460164/src/lib.rs#L99-L103

```rust
    fn block_on<F, Fut>(self, mut handler: F, fut: Fut) -> !
    where
        F: FnMut(Event<'_, Self::User>, &Elwt<Signal<Self::User>>, &mut ControlFlow) + 'static,
        Fut: Future<Output = Infallible> + 'static,
```

**Comment by nerditation** [comment]
> * Document that `RedrawRequested` is no longer guaranteed to be delivered in order with respect to any other event - it is platform specific

> There would no longer be any notion of having multiple redraw events within a single iteration of the loop - you just get (throttled) RedrawRequested events in response to calling request_redraw() - so now you would just submit work to the GPU as part of your RedrawRequested event handler (i.e. after you've finished actually drawing).

@rib can you clarify this please, what is the intended behavior in this new EventLoop 3.0 desing, suppose a `RedrawRequested` is delivered to a event loop that is blocking, will there still be `NewEvent(WaitCanceled)` and `AboutToWait` events emited before and after the `RedrawRequested` respectively? or will there only be a single `RedrawRequested`?

**Comment by rib** [comment]
Ah, that's interesting.

With the initial changes I made for pump events (to the windows backend in particular) my first interest was in consolidating things so we just had one clear place where the backend would get/peek messages and considering the previous complexity that existed for driving redraws it also made sense to drive redraws from a single standard event that at least has some awareness of whether a redraw is needed. I wasn't _really_ looking at the problem of throttling / synchronization at that time.

Later on we had a lot of discussion around how Winit could support throttling of redraw events (mainly for Wayland + Web at the time).

The main change here is that Winit makes it clearer that applications should redraw within RedrawRequested callbacks so then it's at least possible for the backend to throttle the delivery of those RedrawRequested events if the window system provides a suitable mechanism for throttling. (most window systems can find something better than just a timer for throttling).

That doesn't necessarily mean that all backends currently do throttle RedrawRequested events - so far it's really just the Wayland + Web backends where throttling was specifically implemented.

If a backend doesn't throttle these events then, just as with Winit 0.28, applications may end up wasting a lot of energy rendering a lot faster than the compositor can possibly display.

I'm not as familiar with Windows graphics programming as I am Linux, X11/Wayland and although I knew `WM_PAINT` wasn't going to be synchronized with the display/compositor I had got the impression there was at least some crude timer based throttling done by the OS, but it's also not super surprising to learn that it's not throttled at all.

Throttling for the Windows backend can hopefully be implemented based on timing feedback from DWM / the compositor which has been discussed but it's not something that's been looked at yet.

There had been some discussion of having a timer based fallback for throttling too, though I'm not sure if anything like this has been implemented yet. If WM_PAINT is entirely unthrottled though it would potentially make sense to use timer based throttling at least until a better solution based on compositor feedback can be implemented.

**Comment by rib** [comment]
> suppose a RedrawRequested is delivered to a event loop that is blocking, will there still be NewEvent(WaitCanceled) and AboutToWait events emited before and after the RedrawRequested respectively? or will there only be a single RedrawRequested?

more specifically to try and answer this...

In general most applications should ideally avoid making assumptions that these events are related.

Technically, yes NewEvents is going to be called before RedrawRequested, and AboutToWait should be called afterwards but there's no precise relationship guaranteed. NewEvents + AboutToWait may be dispatched many times e.g. due to input or other events and RedrawRequested isn't necessarily going to be called each time the event loop wakes up.

The NewEvents, and AboutToWait events can be thought of as being similar to RunLoop Observer events on macOS in the sense that they give applications some way of observing the internal way that the event loop blocks and waits for new events from the OS, but for all intents and purposes the frequency of this internal OS loop and polling for events is arbitrary and may be completely decoupled from any specific event such as RedrawRequested events.

There could be many factors that would affect the frequency that the underlying even loop wakes up depending on OS specific sources of events that Winit backends are listening for and so NewEvents + AboutToWait may get delivered at a much higher frequency than RedrawRequested events.

Conceptually though the OS event loop has to wake up to dispatch RedrawRequested so in that sense you should expect to see NewEvents come before RedrawRequested and AboutToWait come some time afterwards.

It's perhaps worth noting that it's not technically possible to truly observe the underlying event loop on all platforms (Web in specifically) so it's possible that assumptions about this relationship could be broken between Web and other platforms.

**Comment by madsmtm** [comment]
> is there's a reason it need to be generic

I just discovered this as well, and actually made it independent on macOS and iOS in https://github.com/rust-windowing/winit/pull/3045.

Probably can be done for the remaining platforms (Web, Redox/Orbital and Windows), but it's a bit of work.

**Comment by nerditation** [comment]
> Probably can be done for the remaining platforms (Web, Redox/Orbital and Windows), but it's a bit of work.

that's my impression too. conceptually, imagine we don't support parameterized user events at all (we could, e.g. support an single untyped/erased user event type with dynamic dispatch), the implementation would remain almost unchanged as of today, right? then we can implement paraemterized user events on top of the "bare" event loop, with minimal change required to the code. at least that's my thought process.

however, as I read through some code of the Windows backend, the implementation is intertwined among different parts, although many components never use the generic parameter T, it is still in the signature just to be passed all the way down. indeed some serious work is needed to refactor this.

**Comment by madsmtm** [comment]
I've moved it to https://github.com/rust-windowing/winit/issues/3053

**Comment by Stargateur** [comment]
`process::exit()` make me run away from `winit`. It's so evil to put an exit in a library, panic if you wish but don't exit in place of user.

I see run return Result now but beta doc still state:

> X11 / Wayland: The program terminates with exit code 1 if the display server disconnects.

Do I assume it's still true ?

**Comment by kchibisov** [comment]
@Stargateur are you sane? Have you read what the functions do? Could you, please, try to read into the API, what it does, etc, before commenting like that, thanks.

**Comment by Stargateur** [comment]
@kchibisov I suggest you read my comment twice cause I'm not talking about `ExitWithCode`.

**Comment by kchibisov** [comment]
I suggest you to actually find the `std::process::exit` in the winit codebase.

**Comment by Stargateur** [comment]
doesn't change anything from user side of view, I don't want who call exit but anyway that what winit do by calling whatever that call exit. I know a bit of wayland and wayland-rs and I don't think they call exit either.

**Comment by kchibisov** [comment]
I'm saying that no-one calls `exit`, but you can believe what you believe, since you somehow concluded that winit calls `exit` now.

**Comment by wleslie** [comment]
That said, the [docs for `EventLoop::run` in 0.29.1-beta](https://docs.rs/winit/0.29.1-beta/winit/event_loop/struct.EventLoop.html#method.run) say "X11 / Wayland: The program terminates with exit code 1 if the display server disconnects."

One reasonable interpretation of that sentence is that `winit` calls `std::process::exit` in that situation. Since that's not true, then what is the right way to interpret that sentence? What's responsible for terminating the program with exit code 1, if not `winit`?

**Comment by rib** [comment]
fwiw Winit did use to call `process::exit` in `run` and this was actually one of the specific things I was aiming to address for the Android backend since it was especially problematic that `run()` would avoid returning by using `process::exit()` (re: https://github.com/rust-windowing/winit/issues/2709)

Now that Winit's `run()` API can return a `Result` it no longer needs to call `process::exit()`.

In general the previous use of `process:exit()` was specifically done so that `run()` could be declared to never return (i.e. like `run() -> !`) to (effectively) make every platform consistent with iOS which technically can't return from it's main loop once its started.

**Comment by kchibisov** [comment]
>That said, the [docs for EventLoop::run in 0.29.1-beta](https://docs.rs/winit/0.29.1-beta/winit/event_loop/struct.EventLoop.html#method.run) say "X11 / Wayland: The program terminates with exit code 1 if the display server disconnects."

It could be just forgotten to update, since we don't and we always forward errors now, even from other functions. It's best to remove such references all together if you still see them in the docs. You can open a separate issue and I fix the docs once I have time. 

It's just the claim was that we straight terminate, said on that issue, given that this issue was solely about removing any sort of such termination, which I treat as complete dismissal of what was said before on the issue and they just decided to attack without understanding and reading into the issue and following anything.

**Comment by Stargateur** [comment]
> It's just the claim was that we straight terminate, said on that issue, given that this issue was solely about removing any sort of such termination, which I treat as complete dismissal of what was said before on the issue and they just decided to attack without understanding and reading into the issue and following anything.

I didn't attack on anything, I say I also prefer winit without exit and ask if it was still the doc cause the doc on beta still say there is exit for x11 and wayland. I asked you to read twice my comment and you proved you didn't.

**Comment by kchibisov** [comment]
See this https://github.com/rust-windowing/winit/pull/3111 .

**Comment by kchibisov** [comment]
>process::exit() make me run away from winit. It's so evil to put an exit in a library, panic if you wish but don't exit in place of user.

I'd just say that the first sentence was made me decide to discard and not read much into the other part of message for the following reasons:

1. Winit always had options to run without `std::process::exit` even when said function was used in `run`. The `exit` was justified and explicitly mentioned in the docs (not anymore) and why it was used (mostly ios, and I have no clue how it works now, the control flow will never return to the user of the `run` on ios).
2. The `std::process::exit` is a common thing in the GUI world (look at gtk) and not entirely bad(only `_exit` is bad) to call when you can't really return control back. You can also run your functions on `std::process::exit` and that's how some libraries work (nvidia libs run their cleanup like that, panic = "abort" will prevent them to do so for example).
3. Saying that `panic` is better is not correct, because it won't run any `atexit` hooks when the panic is basically `abort`. And `panic = "abort"` is common when you want to get backtraces from your GUI program unless you redirect and collect the logs. So you'd at least get the `dump` if something failed.

I'm sorry for being offensive, it's just, hard to understand what folks really want when they start their comments with a sentence making me not want to read the rest, and that the thing you said was only half the truth, given that winit **always** offered alternatives which **never** terminated with `exit` or `panic` and normally returned (Old polling stuff, then `run_return`, and now `run` becomes like that). It's hard to understand when ppl just extrapolate their feeling or just trying to offense, since as a maintainer, I get comments like that on daily basis not only in winit, so over the time your brain starts to ignore them and not even making you to read through them.

Just simply linking to the exact place in docs where you have question **is always better** and you could save a lot of time removing all the misunderstanding. I fixed it after @wleslie provided relevant link to the docs.

**Comment by epimeletes** [comment]
Apologies for what may seem like bikeshedding, but `run_ondemand` should really be called `run_on_demand`, "ondemand" is not a word in the English language.

**Comment by rib** [comment]
Yeah, would probably make sense. The name really just came from terms that were being used in the original github issues being addressed (https://github.com/rust-windowing/winit/issues/2431). It became the term we used for consistency to refer to that use case.

I guess I won't get a chance to make a PR to change it myself but it does sound like it would be reasonable to rename.

**Comment by kchibisov** [comment]
fine with renaming, do you want to send a patch doing so?

**Comment by simonask** [comment]
(Not sure if this is the right place to ask, so feel free to hide/delete this comment.)

First off, thanks for all the hard work!

I'm migrating my codebase (a game) to use winit 0.29, and I'm really confused by some of the guidance in the docs, but I'm also unsure how much of it is intentional or lacking design details. Here are some questions that popped up:

1. I noticed that resize behavior changed on Windows (#3272), seemingly due to less magic happening around the resize modal loop, causing `ControlFlow::Poll` to not trigger `AboutToWait` during a resize. This maps well to the platform behavior, but AFAICT there is no longer any way to get continuous updates during the resize modal loop. You can hook into the `Resized` event, but it is only emitted when the window is actually resized (the mouse moved while dragging the corners or the window). Is there a generally suggested workaround?

2. The most obvious workaround is to simply update/draw on a dedicated thread, independent of the event loop. This simplifies some things, but adds machinery to schlep events between threads. But the documentation for `pump_events()` discourages buffering events, because some events on some platforms need to be handled within the callback. This in turn discourages threading as a solution for the problem described in (1), whether or not `pump_events()` is being used or not.

3. Ok fine, so let's handle events on the main thread, and just render on a background thread. But one of the events that `pump_events()` mentions needs to be handled in the callback is `RedrawRequested`, in order to avoid visual tearing. This seems like a much too strong claim for the use case of rendering to a surface using for example `wgpu`, where my assumption would be that the presentation machinery (specifically VSync) would take care of this synchronization. In other words, unless I'm mistaken, the guidance should be to either render in `RedrawRequested` OR ensure some other kind of synchronization, like GPU presentation modes. I could be wrong here, but `wgpu` should be using displayLink internally on macOS/iOS to achieve this, leaving winit out of the picture. Furthermore, it's a really common design in game engines to have a dedicated render thread, that isn't the main thread, which takes care of command buffer submission and presentation.

4. Another potential workaround to (1) is to call `request_redraw()` inside the handling of `RedrawRequested` while resizing. This seems super ugly, and kind of distributes the responsibility of maintaining the desired behavior of `ControlFlow::Poll` to multiple places in the code for a very platform-specific quirk.

Really sorry if I'm missing something obvious. I'm not really sure what is actionable about all this - the changes seem really good overall, and I like that there is less magic going on around the quirks of the Windows event loop. Might just be a documentation issue. :-)

**Comment by kchibisov** [comment]
>. This maps well to the platform behavior, but AFAICT there is no longer any way to get continuous updates during the resize modal loop. You can hook into the Resized event, but it is only emitted when the window is actually resized (the mouse moved while dragging the corners or the window). Is there a generally suggested workaround?

that's a bug that should be fixed given that it's a windows specific issue.

>where my assumption would be that the presentation machinery (specifically VSync) would take care of this synchronization. In other words, unless I'm mistaken, the guidance should be to either render in RedrawRequested OR ensure some other kind of synchronization, like GPU presentation modes. I could be wrong here, but wgpu should be using displayLink internally on macOS/iOS to achieve this, leaving winit out of the picture. Furthermore, it's a really common design in game engines to have a dedicated render thread, that isn't the main thread, which takes care of command buffer submission and presentation.

This is only on macOS since it really wants you to draw right from this event otherwise you may get a bit undesired artifacts, but it's not tearing, not.



>Another potential workaround to (1) is to call request_redraw() inside the handling of RedrawRequested while resizing. This seems super ugly, and kind of distributes the responsibility of maintaining the desired behavior of ControlFlow::Poll to multiple places in the code for a very platform-specific quirk.

Calling redraw_requested from RedrawRequested itself is supported and guaranteed to work on all the platforms. It'll properly send it on the next frame.

>(EDIT) The docs also generally discourage using AboutToWait to do anything useful, and instead using RedrawRequested to do continuous rendering, implying that request_redraw() is called at the end of every frame to schedule the next frame for rendering. But then what is the intended use of ControlFlow::Poll?

To spin the loop as fast as you can? Like not sure why it shouldn't exist given that it's pretty common for event loop to have wait, wait with time out, and instant return when no events.

**Comment by simonask** [comment]
> that's a bug that should be fixed given that it's a windows specific issue.

Thank you for clarifying this!

> This is only on macOS since it really wants you to draw right from this event otherwise you may get a bit undesired artifacts, but it's not tearing, not.

It's still unclear to me what the guidance actually is here. Is it the case for all use cases that any drawing on the window's surface, including presentation to a GPU surface, should happen within `RedrawRequested` on macOS? Maybe this question actually belongs in `wgpu`.

Really sorry if this is hijacking the issue, please let me know if there is a better place to put these questions!

**Comment by kchibisov** [comment]
>It's still unclear to me what the guidance actually is here. Is it the case for all use cases that any drawing on the window's surface, including presentation to a GPU surface, should happen within RedrawRequested on macOS?

it's recommended and how apply expects you.

>Really sorry if this is hijacking the issue, please let me know if there is a better place to put these questions!

you should ask such questions on matrix or discussions.

**Comment by yanchith** [comment]
Hi all! Came here after attempting to upgrade to 29. I am wondering, how should one structure a loop for videogames these days.

In 28, we had `MainEventsCleared` and `RedrawRequested`. I used to do game simulation code in `MainEventsCleared` and then manually fire `window.request_redraw()`. Now that `MainEventsCleared` got removed, where does one put the simulation code?

I notice there's `AboutToWait`, but the docs say I shouldn't need to use it?

> Most applications shouldn’t need to hook into this event since there is no real relationship between how often the event loop needs to wake up and the dispatching of any specific events.

Thanks!

**Comment by daxpedda** [comment]
If you want to draw or update in VSync, you can do that inside `WindowEvent::RedrawRequested` and call `Window::request_redraw()` (see #3406) there as well.

If you want to draw or update in a specific interval, you can use `ControlFlow::WaitUntil` to schedule that interval, and either do your thing in `Event::NewEvents` or `Event::AboutToWait`.

If you want to draw or update as fast possible, which is not recommended as it will just consume 100% CPU, you can use `ControlFlow::Poll` and do your thing in `Event::NewEvents` or `Event::AboutToWait` as well.

**Comment by JoeOsborn** [comment]
This contradicts the guidance of https://docs.rs/winit/latest/winit/event_loop/enum.ControlFlow.html , and I guess in WGPU at least waiting for the swapchain image can block which is how vsync throttling typically is done for that API.  Is your note here about the case where you don't have a 3D API, or are the docs wrong here?

**Comment by daxpedda** [comment]
The documentation is poor and wrong in some places.

Unfortunately there are some inconsistencies here, unfortunately Wgpu doesn't always block for VSync, e.g. Web doesn't, so if you use `ControlFlow::Poll` on Web with Wgpu it will just break your application (throttle it into oblivion).

The fact is that there is no ideal solution here until either Wgpu improves (https://github.com/gfx-rs/wgpu/issues/2869, https://github.com/gfx-rs/wgpu/issues/3283) or Winit finishes #2412.

But if you have a reliable method blocking until VSync and you don't want to handle any events while doing that, `ControlFlow::Poll` can work fine as well.

**Comment by kchibisov** [comment]
Would close it since it basically a meta issue at this point. Specific issues that still are present with event loop has their corresponding issues pretty much.
